src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -41,6 +41,20 @@
 #define TRANSIENT_FAILURE_ATTEMPTS 3 /* Arbitrary limit: how many attempts are considered enough to report
                                       * transient failure. */
 
+/* RFC2131 section 4.1 specifies retransmission delays, starting at 4s then doubling
+ * each attempt with max of 64s, with -1 to +1 sec of random 'fuzz' added */
+#define RFC2131_FUZZ_USEC ((int32_t)(random_u32() % (2 * USEC_PER_SEC)) - (int32_t)USEC_PER_SEC)
+#define RFC2131_SEC41_NEXT_TIMEOUT_USEC(now, attempt) \
+        ((usec_t)(now + ((int32_t)((UINT64_C(1) << MIN((attempt) + 1, (uint64_t) 6)) * USEC_PER_SEC) + RFC2131_FUZZ_USEC)))
+
+/* RFC2131 section 4.4.5 specifies T1 (renewing) and T2 (rebinding) defaults and
+ * special reacquisition delay algorithm, different from above section 4.1,
+ * however we use the same -1 to +1 sec 'fuzz' from above */
+#define RFC2131_T1_DEFAULT(lifetime) ((lifetime) / 2)
+#define RFC2131_T2_DEFAULT(lifetime) (((lifetime) * 7) / 8)
+#define RFC2131_SEC445_NEXT_TIMEOUT_USEC(now, expire) \
+        ((usec_t)(now + (MAX((((int32_t)((expire) - (now)) / 2) + RFC2131_FUZZ_USEC), (int32_t)(60 * USEC_PER_SEC)))))
```

Comment:
as above

Reply:
why the downcase to 32bit btw?

Reply:
i was casting to signed int to handle the situation where expire <= now, though it should have been int64_t; however in the latest push i just changed it to expire > now ?: which is probably clearer, and removed the signed casting. let me know if you prefer a different way, or if you think it might be better to move the logic into a function instead of a macro.